packages/plugin-workspace-tools/sources/commands/foreach.ts

```diff
@@ -126,14 +134,30 @@ export default class WorkspacesForeachCommand extends BaseCommand {
       ? project.topLevelWorkspace
       : cwdWorkspace!;
 
+    if (configuration.projectCwd === null)
+      throw new UsageError(`This command can only be run from within a Yarn project`);
+
+    const root = this.since ? await gitUtils.fetchRoot(configuration.projectCwd) : null;
+    const base = this.since && root !== null
+      ? await gitUtils.fetchBase(root, {baseRefs: typeof this.since === `string` ? [this.since] : configuration.get(`changesetBaseRefs`)})
+      : null;
+
+    const prospectiveWorkspaces = this.since && root !== null
+      ? Array.from(await gitUtils.fetchChangedWorkspaces(root, {base: base!.hash, project}))
+      : [rootWorkspace, ...(this.from.length > 0 ? rootWorkspace.getRecursiveWorkspaceChildren() : [])];
+
     const fromPredicate = (workspace: Workspace) => micromatch.isMatch(structUtils.stringifyIdent(workspace.locator), this.from);
     const fromCandidates: Array<Workspace> = this.from.length > 0
-      ? [rootWorkspace, ...rootWorkspace.getRecursiveWorkspaceChildren()].filter(fromPredicate)
-      : [rootWorkspace];
-
-    const candidates = this.recursive
-      ? [...fromCandidates, ...fromCandidates.map(candidate => [...candidate.getRecursiveWorkspaceDependencies()]).flat()]
-      : [...fromCandidates, ...fromCandidates.map(candidate => [...candidate.getRecursiveWorkspaceChildren()]).flat()];
+      ? prospectiveWorkspaces.filter(fromPredicate)
+      : prospectiveWorkspaces;
+
+    const candidates = new Set([...fromCandidates, ...(await Promise.all(fromCandidates.map(async candidate => [...(
+      this.recursive
+        ? this.since
+          ? await candidate.getRecursiveWorkspaceDependents()
+          : candidate.getRecursiveWorkspaceDependencies()
```

Comment:
These three lines precisely illustrate what I was trying to call out in https://github.com/yarnpkg/berry/issues/3450:

> I think the weirdness here stems partly from `--recursive`, which in the current implementation of `foreach` acts like an exclusionary filter, not an inclusionary one. This behavior is reversed in the presence of `--since`.

A flag doing opposite things in the presence of other flags is not unheard of, but it's certainly unusual. You could explain this away by saying that `--since` is "earlier" or "higher level" or "coarser" than `--recursive`. In other words: while you can take an invocation of the command that uses `--since` and refine it (it some sense) with `--recursive`, the opposite doesn't really make sense. You can't _refine_ an invocation using `--recursive` by adding `--since` because really, it broadens the selection.

I threw around some half-baked ideas in that issue for a holistic re-evaluation of how the various flags of `foreach` should interact, but that's probably out of scope for this PR.

Reply:
Yeah I definitely agreed with that point on reading your proposal. That's why I haven't fully implemented it at this stage as I wanted to confirm whether it is appropriate to use `--recursive` for this, or if it warrants its own option.

If everyone is happy to use `--recursive` here, then I will be sure to update the command description to make this behaviour clear, as well as implementing the same behaviour for `yarn workspaces list`.